Grammar and formatting revisions for snapshot reads spec (#1037)
mongodb/specifications: source/sessions/snapshot-sessions.rst

Snapshot Reads Specification
============================

:Spec Title: Snapshot Reads Specification
:Spec Version: 1.2.1
:Author: Boris Dogadov
:Advisors: Jeff Yemin, A. Jesse Jiryu Davis, Judah Schvimer
:Status: Accepted
:Type: Standards
:Minimum Server Version: 5.0

> :Last Modified: 12-Jul-2021

Author: ```suggestion
:Last Modified: 15-Jul-2021
```

> :Last Modified: 15-Jul-2021

Author: ```suggestion
:Last Modified: 16-Jul-2021
```

    Snapshots reads are majority committed reads.

Snapshot timestamp
    Snapshot timestamp, representing timestamp of the first supported read operation (i.e. find/aggregate/distinct) in the session.
    The server creates a cursor in response to a snapshot find/aggregate command and 

>     reports ``atClusterTime`` within the ``cursor`` field in the response. For distinct commands the server adds a top-level ``atClusterTime`` field to the response.

Reviewer: "for the distinct command the server" to match "find/aggregate command" singular form?

Author: ```suggestion
    reports ``atClusterTime`` within the ``cursor`` field in the response. For the distinct command the server adds a top-level ``atClusterTime`` field to the response.
```

If no value is provided for ``snapshot`` a value of false is
implied.
There are no MongoDatabase, MongoClient, or MongoCollection API changes.

> SessionOptions Changes

Author: ```suggestion
SessionOptions changes
```

this property is false.

Snapshot reads and causal consistency are mutually exclusive. Therefore if ``snapshot`` is set to true,
``causalConsistency`` must be false. Client MUST throw an error if both ``snapshot`` and ``causalConsistency`` are set to true.
Snapshot reads are supported on both primaries and secondaries.

> ClientSession Changes

Author: ```suggestion
ClientSession changes
```

=====================

> Transaction are not allowed with snapshot sessions.

Reviewer 2: ```suggestion
Transactions are not allowed with snapshot sessions.
```

> Calling ``session.startTransaction(options)`` on a snapshot session SHOULD raise an error.

Author: Not sure why the original text wasn't using "MUST" but I opted just to capitalize the term to avoid any functional changes with this PR.

Reviewer 3: Do you mind changing this to "MUST"?

Author: ```suggestion
Calling ``session.startTransaction(options)`` on a snapshot session MUST raise an error.
```

> ReadConcern Changes

Reviewer: Should the case match the case used in [sessions spec](https://github.com/mongodb/specifications/blob/master/source/sessions/driver-sessions.rst) ?

Author: I'll revert this. IMO, there are wild inconsistencies across our specs when it comes to headings and [title capitalization](https://en.wikipedia.org/wiki/Title_case).

Author: ```suggestion
ReadConcern changes
```

        atClusterTime : <BsonTimestamp>
    }

The ``atClusterTime`` timestamp MUST be stored in the ``ClientSession`` to later be passed as the

> ``atClusterTime`` field of a ``readConcern`` with a ``snapshot`` level in subsequent read operations.

Reviewer 2: I actually think `the readConcern` reads better than `a readConcern` since the readConcern is not optional for these operations

Author: ```suggestion
``atClusterTime`` field of the ``readConcern`` with a ``snapshot`` level in subsequent read operations.
```

Server Errors
=============
1. The server may reply to read commands with a ``SnapshotTooOld(239)`` error if the client's ``atClusterTime`` value is not available in the server's history.
2. The server will return ``InvalidOptions(72)`` error if both ``atClusterTime`` and ``afterClusterTime`` options are set to true.

> 3. The server will return ``InvalidOptions(72)`` error if the command does not support readConcern.level "snapshot".

Author: Pulled these numeric codes from https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml. I don't think the names are very useful on their own, since driver tests would ideally check for codes.

Reviewer 3: SGTM.

This goal is met by defining a ``SessionOptions`` field that applications use to
start a ``ClientSession`` that can be used for snapshot reads. Alternative explicit approach of
obtaining ``atClusterTime`` from ``cursor`` object and passing it to read concern object was considered initially.
A session-based approach was chosen as it aligns better with the existing API, and requires minimal API changes.
Future extensibility for snapshot reads would be best served by a session-based approach, as no API changes will be required.

Backwards Compatibility
=======================

> The API changes to support snapshot reads extends the existing API but does not

Author: I assumed "to support sessions" was copypasta from the other session spec.

Reviewer: copypasta indeed.
Should it stay "extend" and "do"?

Author: ```suggestion
The API changes to support snapshot reads extend the existing API but do not
```

:2021-06-15: Initial version.
:2021-06-28: Raise client side error on < 5.0.
:2021-06-29: Send readConcern with all snapshot session commands.

> :2021-07-12: Grammar and formatting revisions

Author: ```suggestion
:2021-07-12: Grammar revisions. Change SHOULD to MUST for startTransaction error to comply with existing tests.
```

> :2021-07-15: Grammar revisions. Change SHOULD to MUST for startTransaction error to comply with existing tests.

Author: ```suggestion
:2021-07-16: Grammar revisions. Change SHOULD to MUST for startTransaction error to comply with existing tests.
```
